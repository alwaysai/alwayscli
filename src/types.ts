import { BRANCH, LEAF } from './constants';

type Argv<TRequired extends boolean> = TRequired extends true
  ? string[]
  : (string[] | undefined);

export type CliInput<TValue, TRequired extends boolean = boolean> = {
  placeholder: string;
  getValue:
    | ((argv: Argv<TRequired>) => TValue)
    | ((argv: Argv<TRequired>) => Promise<TValue>);
  description?: string;
  required?: TRequired;
  hidden?: boolean;
};

export type AnyInput = CliInput<any>;

export type InputValue<TInput> = TInput extends CliInput<infer U, any> ? U : never;

export type AnyNamedInputs = {
  [name: string]: AnyInput;
};

export type NamedValues<TNamedInputs extends AnyNamedInputs> = {
  [K in keyof TNamedInputs]: InputValue<TNamedInputs[K]>
};

export type Command = CliBranch | CliLeaf<AnyInput, AnyNamedInputs, AnyInput>;

export type CliBranch = {
  commandType: typeof BRANCH;
  name: string;
  description?: string;
  hidden?: boolean;
  subcommands: (CliBranch | CliLeaf<any, any, any>)[];
  next?: CliBranch | CliLeaf<any, any, any>;
};

export type CliLeaf<
  TPositional extends AnyInput,
  TNamed extends AnyNamedInputs,
  TEscaped extends AnyInput
> = {
  commandType: typeof LEAF;
  name: string;
  action: (
    positionalValue: InputValue<TPositional>,
    namedValues: NamedValues<TNamed>,
    escapedValue: InputValue<TEscaped>,
  ) => any;
  positionalInput?: TPositional;
  namedInputs?: TNamed;
  escapedInput?: TEscaped;
  description?: string;
  hidden?: boolean;
};

// The "commandType" field is assigned internally by the framework.
// This helper function is used to remove that field for the input
// type of the createLeaf and CliBranch factories.
export type ExcludeCommandType<T extends { commandType: any }> = Pick<
  T,
  Exclude<keyof T, 'commandType'>
>;
