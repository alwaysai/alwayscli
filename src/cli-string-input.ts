import { CliInput } from './types';
import { CliUsageError } from './cli-usage-error';

type Config = Partial<{
  required: boolean;
  description: string;
  defaultValue: string;
  hidden?: boolean;
  placeholder?: string;
}>;

export { CliStringInput };
function CliStringInput(
  config: Config & { defaultValue: string },
): CliInput<string, false>;
function CliStringInput(config: Config & { required: true }): CliInput<string, true>;
function CliStringInput(config?: Config): CliInput<string | undefined, false>;
function CliStringInput(config: Config = {}) {
  const {
    defaultValue,
    required = false,
    description,
    placeholder = '<str>',
    hidden = false,
  } = config;
  const input: CliInput<string | undefined> = {
    hidden,
    placeholder,
    required,
    getValue(argv) {
      if (!argv) {
        return typeof defaultValue === 'string' ? defaultValue : undefined;
      }

      if (argv.length > 1) {
        throw new CliUsageError(`Expected just one ${placeholder}`);
      }

      if (argv.length === 0) {
        throw new CliUsageError(`Expected a ${placeholder}`);
      }

      return argv[0];
    },
    description,
  };
  return input;
}
