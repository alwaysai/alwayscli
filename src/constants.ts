import chalk from 'chalk';

export const LEAF = 'leaf';
export const BRANCH = 'branch';
export const RED_ERROR = chalk.red('Error:');
