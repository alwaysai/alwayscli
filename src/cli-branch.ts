import { CliBranch, ExcludeCommandType } from './types';
import { BRANCH } from './constants';

type Config = ExcludeCommandType<CliBranch>;

export function CliBranch(config: Config) {
  const branch: CliBranch = {
    ...config,
    commandType: BRANCH,
  };
  return branch;
}
