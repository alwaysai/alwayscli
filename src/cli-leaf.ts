import { AnyNamedInputs, CliLeaf, ExcludeCommandType, AnyInput, CliInput } from './types';
import { LEAF } from './constants';

export function CliLeaf<
  TPositional extends AnyInput = CliInput<undefined, false>,
  TNamed extends AnyNamedInputs = {},
  TEscaped extends AnyInput = CliInput<undefined, false>
>(config: ExcludeCommandType<CliLeaf<TPositional, TNamed, TEscaped>>) {
  const cliLeaf: CliLeaf<TPositional, TNamed, TEscaped> = {
    ...config,
    commandType: LEAF,
  };
  return cliLeaf;
}
