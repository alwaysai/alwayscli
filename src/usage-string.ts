import redent = require('redent');

import { Command, AnyInput } from './types';
import { CLI_BRANCH, RED_ERROR } from './constants';
import { createTextList } from './create-text-list';
import { regularizeText, wrapInSquareBrackets } from './util';
import { getPathAndDescriptionOfLeaves } from './get-path-and-description-of-leaves';
import { LastCommand } from './last-command';
import { mapCommand } from './map-command';

const INDENT_SIZE = 3;

export function UsageString(rootCommand: Command, errorMessage?: string) {
  const lastCommand = LastCommand(rootCommand);

  const commandPathString = mapCommand(rootCommand, command => command.name).join(' ');
  let firstParagraph = `Usage: ${commandPathString}`;
  const otherParagraphs: string[] = [];

  function appendInputUsage(input?: AnyInput, prefix?: string) {
    if (input && !input.hidden) {
      const { placeholder, description, required } = input;
      if (prefix) {
        firstParagraph += ` ${prefix}`;
      }
      firstParagraph += ` ${required ? placeholder : wrapInSquareBrackets(placeholder)}`;
      if (description) {
        otherParagraphs.push(`${placeholder}:`);
        otherParagraphs.push(redent(regularizeText(description), INDENT_SIZE));
      }
    }
  }

  if (lastCommand.commandType === CLI_BRANCH) {
    // BRANCH
    firstParagraph += ' <subcommand> ...';
    otherParagraphs.push('Subcommands:');
    const nameAndDescriptionOfLeaves = getPathAndDescriptionOfLeaves(lastCommand, []);
    const items: Parameters<typeof createTextList> = nameAndDescriptionOfLeaves.map(
      ({ path, description }) => ({
        name: path.join(' '),
        text: description,
      }),
    );
    const subcommandsParagraph = redent(createTextList(...items), INDENT_SIZE);
    otherParagraphs.push(subcommandsParagraph);
  } else {
    // LEAF
    const { positionalInput, namedInputs, escapedInput } = lastCommand;

    appendInputUsage(positionalInput);

    if (namedInputs) {
      const entries = Object.entries(namedInputs).filter(([_, input]) => !input.hidden);
      if (entries.length > 0) {
        const optionsNotRequired = entries.every(
          ([_, namedInput]) => !namedInput.required,
        );
        firstParagraph += optionsNotRequired ? ' [<options>]' : ' <options>';
        otherParagraphs.push('Options:');
        const items: Parameters<typeof createTextList> = entries.map(
          ([inputName, input]) => {
            let name = `--${inputName}`;
            if (input.placeholder) {
              name += ` ${input.placeholder}`;
            }
            if (!input.required) {
              name = wrapInSquareBrackets(name);
            }
            const text = input.description;
            return { name, text };
          },
        );
        const optionsParagraph = redent(createTextList(...items), INDENT_SIZE);
        otherParagraphs.push(optionsParagraph);
      }
    }

    appendInputUsage(escapedInput, '--');
  }

  const paragraphs = [firstParagraph];

  const regularizedDescription = regularizeText(lastCommand.description);
  if (regularizedDescription) {
    paragraphs.push(redent(regularizedDescription, INDENT_SIZE));
  }

  paragraphs.push(...otherParagraphs);

  if (errorMessage) {
    paragraphs.push(`${RED_ERROR} ${errorMessage}`);
  }

  return paragraphs.join('\n\n');
}
