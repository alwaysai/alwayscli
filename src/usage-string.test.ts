import { UsageString } from './usage-string';
import { CliBranch } from './cli-branch';
import { CliStringInput } from './cli-string-input';
import { CliLeaf } from './cli-leaf';
import { Command } from './types';
import { RED_ERROR } from './constants';

const messageInput = CliStringInput({ description: 'A string message please' });
const positionalInput = CliStringInput({ description: 'A word', placeholder: '<word>' });
const escapedInput = CliStringInput({ description: 'Another word', required: true });

const leaf = CliLeaf({
  name: 'echo',
  positionalInput,
  namedInputs: {
    message: messageInput,
  },
  escapedInput,
  action(foo) {
    return foo;
  },
});

const root = CliBranch({
  name: 'cli',
  description: 'This is a CLI',
  subcommands: [leaf],
});

describe(UsageString.name, () => {
  it('Creates a usage string for a branch', () => {
    root.next = undefined;
    const usageString = UsageString(root);
    expect(usageString).toMatchSnapshot();
  });

  it('Creates a usage string for a leaf', () => {
    root.next = leaf;
    const usageString = UsageString(root);
    expect(usageString).toMatch(messageInput.description!);
    expect(usageString).toMatchSnapshot();
  });

  it('Creates a usage string for a leaf without a parent branch', () => {
    const usageString = UsageString(leaf as Command);
    expect(usageString).toMatchSnapshot();
  });

  it('Writes an error message at the end if one is provided', () => {
    const usageString = UsageString(root, 'foo');
    expect(usageString.endsWith(`${RED_ERROR} foo`)).toBe(true);
  });

  it('Does not write usage for named inputs if there are none', () => {
    const usageString = UsageString(CliLeaf({ name: 'foo', action() {} }));
    expect(UsageString(CliLeaf({ name: 'foo', action() {}, namedInputs: {} }))).toBe(
      usageString,
    );
    expect(usageString).toMatchSnapshot();
  });
});
