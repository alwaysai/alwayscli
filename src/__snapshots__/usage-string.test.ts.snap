// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsageString Creates a usage string for a branch 1`] = `
"Usage: cli <subcommand> ...

   This is a CLI

Subcommands:

   echo"
`;

exports[`UsageString Creates a usage string for a leaf 1`] = `
"Usage: cli echo [<word>] [<options>] -- <str>

<word>:

   A word

Options:

   [--message <str>] : A string message please

<str>:

   Another word"
`;

exports[`UsageString Creates a usage string for a leaf without a parent branch 1`] = `
"Usage: echo [<word>] [<options>] -- <str>

<word>:

   A word

Options:

   [--message <str>] : A string message please

<str>:

   Another word"
`;

exports[`UsageString Does not write usage for named inputs if there are none 1`] = `"Usage: foo"`;
